File: python/8queens.py
#!/usr/bin/python3

N = 8
queens = [0] * N

def check(x1, y1, x2, y2):
  if y1 == y2:
    return False
  if x2 - x1 == y2 - y1 or x2 - x1 == y1 - y2:
    return False
  return True

def find(q):
  if q == N:
    for i in range(N):
      pos = queens[i]
      line = '.' * pos + 'Q' + '.' * (N - pos - 1)
      print(line)
    print()
    return
  for i in range(N):
    ok = True
    for k in range(q):
      ok = check(k, queens[k], q, i)
      if not ok:
        break
    if ok:
      queens[q] = i
      find(q + 1)

find(0)

File: cpp/8queens.cpp
#include <iostream>
#include <string>

const int N = 8;

int queens[N];

bool check(int x1, int y1, int x2, int y2)
{
  if (y1 == y2)
  {
    return false;
  }
  if (x2 - x1 == y2 - y1 || x2 - x1 == y1 - y2)
  {
    return false;
  }
  return true;
}

void find(int q)
{
  if (q == N)
  {
    for (int i = 0 ; i < N ; ++i)
    {
      std::string line(N, '.');
      line[queens[i]] = 'Q';
      std::cout << line << '\n';
    }
    std::cout << '\n';
    return;
  }
  for (int i = 0 ; i < N ; ++i)
  {
    bool ok = true;
    for (int k = 0 ; ok && k < q ; ++k)
    {
      ok = check(k, queens[k], q, i);
    }
    if (ok)
    {
      queens[q] = i;
      find(q + 1);
    }
  }
}

int main()
{
  find(0);
}
